src/compiler/file_uploader/index.ts

```diff
@@ -0,0 +1,187 @@
+import { Actor, ActorSubclass, HttpAgent } from '@dfinity/agent';
+import { Secp256k1KeyIdentity } from '@dfinity/identity-secp256k1';
+import { existsSync, createReadStream } from 'fs';
+import { readFile } from 'fs/promises';
+import {
+    getCanisterId,
+    getIdentityName,
+    getWebServerPort
+} from '../../../test';
+import { readdir, stat } from 'fs/promises';
+import { homedir } from 'os';
+import { join } from 'path';
+
+type Src = string;
+type Dest = string;
+
+export async function uploadFiles(canisterName: string, paths: [Src, Dest][]) {
+    const actor = await createUploadFileChunkActor(
+        getCanisterId(canisterName),
+        getWebServerPort(),
+        getIdentityName()
+    );
+
+    const chunkSize = 2_000_000; // The current message limit is about 2 MiB
+
+    for (const [srcPath, destPath] of paths) {
+        // Await each upload so the canister doesn't get overwhelmed by requests
+        await upload(srcPath, destPath, chunkSize, actor);
+    }
+}
+
+async function upload(
+    srcPath: Src,
+    destPath: Dest,
+    chunkSize: number,
+    actor: ActorSubclass
+) {
+    if (!existsSync(srcPath)) {
+        throw new Error(`${srcPath} does not exist`);
+    }
+
+    const stats = await stat(srcPath);
+    if (stats.isDirectory()) {
+        // Await each upload so the canister doesn't get overwhelmed by requests
+        await uploadDirectory(srcPath, destPath, chunkSize, actor);
+    } else {
+        // Await each upload so the canister doesn't get overwhelmed by requests
+        await uploadFile(srcPath, destPath, chunkSize, actor);
+    }
+}
+
+async function uploadDirectory(
+    srcDir: string,
+    destDir: string,
+    chunkSize: number,
+    actor: ActorSubclass
+) {
+    try {
+        const contents = await readdir(srcDir);
+        for (const name of contents) {
+            const srcPath = join(srcDir, name);
+            const destPath = join(destDir, name);
+            // Await each upload so the canister doesn't get overwhelmed by requests
+            await upload(srcPath, destPath, chunkSize, actor);
+        }
+    } catch (error) {
+        console.error(`Error reading directory: ${error}`);
+    }
+}
+
+async function uploadFile(
+    srcPath: Src,
+    destPath: Dest,
+    chunkSize: number,
+    actor: ActorSubclass
+) {
+    console.info(`uploadFile: Uploading ${srcPath} to ${destPath}`);
```

Comment:
I want to see what these logs look like

Reply:
like what?

Reply:
user side "Uploading chunk:" canister sided "Received chunk:"
get rid of start and finish messages.
just hashing: asset/path | start at